packages/sites/src/create-sites-list-component.ts

```diff
@@ -0,0 +1,120 @@
+/* eslint-disable @typescript-eslint/ban-types, react-hooks/rules-of-hooks */
```

Comment:
Looping @Automattic/team-calypso-frameworks to get a second opinion on how to achieve this.

Goals:

1. Allow specifying which hooks should be executed;
2. Reflect types as such (i.e. if `grouping` is disabled, do not expose a `statuses` property in the render child and do not allow passing a `grouping` property specifying grouping options in the usage).

Example:

```tsx
const SitesList = createSitesListComponent({ grouping: false })

<SitesList
  grouping={{ ... }} // this should error
>
  {({ sites, statuses /* this should error */ }) => sites}
</SitesList>
```

If `grouping` is `true`, then:

```tsx
const SitesList = createSitesListComponent({ grouping: false })

<SitesList
  grouping={{ ... }} // this should pass
>
  {({ sites, statuses /* this should pass */ }) => sites}
</SitesList>
```